README.md
This is a refactor exercise and his goal is to overall improve the proposed app. It shows a list of players with data about their team and their league score and we want you to propose solutions and implement them in order to improve the architecture, testability and UX. It's intentionally crappy so you can shine bright :)

## Requirements

There are some mandatory constraints you have to attend:

-   Update the management of the state using REDUX and use a library for the management of the requests (like redux thunk or sagas)
-   Use our component library Ola to build the interface [https://github.com/marketgoo/Ola](https://github.com/marketgoo/Ola)
-   Improve how the user interacts with the form in the way you like. Feel free to add improvements.
-   Add a testing library of your choice and use it to test what you think it's relevant.

## Notes

-   The app contains an api that allows you to get the players, add new ones and delete them.
-   The score of every player is updated every 5 seconds.
-   A socket channel emits a message with the players and their updated scores.
-   Any request to the api will always take an average of 2 seconds to respond.
-   You can install any extra library that is needed without making api changes. (ex: Postcss libraries or for testing)

## Install

-   Donwload repository
-   ~~npm i ( node v12 required )~~

-   `curl https://get.volta.sh | bash`

    > Install Volta from https://volta.sh/ <br />
    > Don't worry about the required node version in this project. Volta select the required version automatically.

    > **Advantage:** Every developer will uses the same required version of node in this project without changes.
    > In others proyects Volta will use other versions of node if it is required.
    >
    > Previously... we had forced the node version in the project with:
    >
    > ```bash
    > $ volta pin node@12
    > success: pinned node@12.22.12 (with npm@6.14.16) in package.json
    > ```

-   `npm ci`

    > Use `npm ci` not `npm i` when its possible.
    >
    > **Advantage:** New minor updates may break the app :-/ Use `ci` instead of `i`. All updates should be controlled & revised by the team.
    > New minor updates may break the app! :-/

## Run app locally

-   webpack -p
-   npm start
-   go to localhost:3000

## API REST

-   **GET /players** return all players saved

```json
{
    "status": "ok",
    "data": [
        {
            "id": 1,
            "name": "Peter",
            "team": "Pandas",
            "score": 5,
            "createdAt": "2020-05-28 11:32:54.506 +00:00",
            "updatedAt": "2020-05-28 11:32:54.506 +00:00"
        },
        {
            "id": 2,
            "name": "Jhon",
            "team": "Cobrakay",
            "score": 8,
            "createdAt": "2020-05-28 11:32:54.507 +00:00",
            "updatedAt": "2020-05-28 11:32:54.507 +00:00"
        }
    ]
}
```

-   **POST /players** Save player data and return player saved

```json
{
    "status": "ok",
    "data": {
        "name": "saved-name",
        "team": "saved-team",
        "score": "saved-score"
    }
}
```

-   **DELETE /players/{id}** Delete a player.

```json
{
    "status": "ok"
}
```


CHANGELOG.md
# Changelog

# Goals sort by priority

(+ priority)

-   Meet the requirements (Redux + Saga + Ola DS) (mandatory!)
-   Fix security issues (mandatory!)
-   Increase robustness (with testing, clean code, apply programming patterns, etc.)
-   UX: Improve the user experience (accessibility, performance, i18n, etc.)
-   Document as much as possible
-   Add new functionalities not proposed (filters, sorting, etc.) (optional)

(- priority)

## Summary

-   [x] Pine Node version with Volta
-   [ ] Add TypeDoc to document code
-   [ ] Redux + sagas
-   [x] Use Ola DS
-   [x] Eslint
-   [ ] Migration to TypeScript
-   [ ] Add typings
-   [x] Add prettier
-   [ ] Add Responsive
-   [x] Add Jest
-   [x] Add React Testing Library
-   [x] Unit Tests
-   [x] Add Cypress.io (e2e tests)
-   [x] Add e2e tests
-   [x] Add react-i18next (i18n)
-   [ ] Refact: Big component to small components
-   [ ] Refact: From Class-Components to Function-Components
-   [ ] Refact: Presentation pattern & Render pattern
-   [ ] Refact: Add react-query (query caching)
-   [ ] Add throttling to the API calls (less requests to the server)
-   [ ] Update React, Webpack, Node & other libraries
-   [ ] Create a docker image to run the app

## Pin node version with Volta

https://volta.sh/

-   Installation: `curl https://get.volta.sh | bash`

-   Pin node version in the project:

    ```bash
    $ volta pin node@12
    success: pinned node@12.22.12 (with npm@6.14.16) in package.json
    ```

## Install Jest

```bash
npm install jest@26
```

New versions doesn't work with Node 12.x. Current version is 29.x.

### Support JSX in Jest

```bash
npm install --save-dev @babel/preset-react @babel/preset-env @babel/plugin-transform-runtime
```

Create .babelrc file with and other presets & plugins:

```json
{
    "presets": ["@babel/preset-env", "@babel/preset-react"],
    "plugins": ["@babel/plugin-transform-runtime"]
}
```

More info:

-   https://babeljs.io/docs/en/babel-preset-env
-   https://babeljs.io/docs/en/babel-preset-react

> Why?
> Transform JSX to JS.
> More info: https://babeljs.io/docs/en/babel-plugin-transform-react-jsx

-   https://babeljs.io/docs/en/babel-plugin-transform-runtime

> Why?
> Babel uses very small helpers for common functions such as \_extend. By default this will be added to every file that requires it. This duplication is sometimes unnecessary, especially when your application is spread out over multiple files.
>
> This is where the @babel/plugin-transform-runtime plugin comes in: all of the helpers will reference the module @babel/runtime to avoid duplication across your compiled output. The runtime will be compiled into your build.

## React Testing Library

```bash
npm install --save-dev @testing-library/react@release-12.x
```

Last versions of React Testing Library doesn't support React 16.x.
We are forced to use an old version (currently).
More info: https://stackoverflow.com/questions/71713405/cannot-find-module-react-dom-client-from-node-modules-testing-library-react#comment127543807_71716438

Other dependencies:

```bash
npm i -D @testing-library/jest-dom
```

## Instal nock (mocking HTTP requests)

https://www.npmjs.com/package/nock

We can directly mock axios with Jest but we make our test and our code will be even more dependent on Axios. In the future we might want to switch from Axios to other request library.

```bash
npm i -D nock
```

## Install [Cypress.io](https://www.cypress.io/) (e2e tests)

```bash
npm install cypress --save-dev
```

Create `.github/workflows/test.yml` to run the tests in the CI (github).

Install [Cypress ESLint plugin](https://github.com/cypress-io/eslint-plugin-cypress) because Cypress uses global variables and we need to tell ESLint to ignore them:

```bash
npm install eslint-plugin-cypress --save-dev
```

```yaml
on: [push]
jobs:
    cypress:
        runs-on: ubuntu-latest
        steps:
            - name: Checkout
              uses: actions/checkout@v3
            - name: Cypress run
              uses: cypress-io/github-action@v5
              with:
                  build: npm run build
                  start: npm start
```

To execute cypress locally (open the cypress UI):

```bash
npm run cypress:open
```

To run Cypress for CI (headless mode):

```bash
npm run cypress:run
```

## Install ESLint

```
npm init @eslint/config
```

Some extra configs needed (jsx support). More info: https://eslint.org/docs/latest/use/configure/

## Install Prettier

https://prettier.io/docs/en/install.html

```bash
npm install --save-dev --save-exact prettier
# Create an empty config file
echo {}> .prettierrc.json
```

There are other options. I use prettier (for now) because it's simple and (almost) standard. But I could use any other code formatter.

## Install Ola DS

```bash
npm i @marketgoo/ola
```

## Migration to TypeScript

> Webpack integration is pretty simple. We are using `ts-loader`, a TypeScript loader, combined with source-map-loader for easier debugging.

https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html#handbook-content

Simply run:

```bash
npm install --save-dev typescript ts-loader
```

And add some changes to the webpack.config.js.

Update webpack, webpack-cli, plugins, etc.

Install types for Jest:

> ts-jest major version must match jest major version

```bash
npm i -D ts-jest@26.4.4 @types/jest
```


cypress/e2e/list.cy.js
/// <reference types="cypress" />

// This spec file contains a variety of tests for the List app.

// The tests are written using the Cypress API.
// https://on.cypress.io/api

describe("List app", () => {
    beforeEach(() => {
        // Intercept HTTP requests
        cy.intercept("/players", {
            status: "ok",
            data: [
                {
                    id: 2,
                    name: "Jhon",
                    team: "Cobrakay",
                    score: 1270,
                    createdAt: "2023-01-29 17:02:43.831 +00:00",
                    updatedAt: "2023-01-29 17:33:00.004 +00:00",
                },
                {
                    id: 1,
                    name: "Peter",
                    team: "Pandas",
                    score: 1129,
                    createdAt: "2023-01-29 17:02:43.831 +00:00",
                    updatedAt: "2023-01-29 17:31:50.004 +00:00",
                },
            ],
        }).as("getPlayers");
        cy.visit("http://localhost:3000");
    });

    it("displays two players with player names when the API response has two players", () => {
        cy.wait("@getPlayers");
        cy.get("[data-testid=player]").should("have.length", 2);

        cy.get("[data-testid=player]").first().should("include.text", "Jhon");
        cy.get("[data-testid=player]").last().should("include.text", "Peter");
    });

    it("can add new players", () => {
        const name = "Dutch";
        cy.intercept("GET", "/players").as("getPlayers");

        cy.intercept("POST", "/players", {
            status: "ok",
            data: { name: name, team: null, score: null },
        }).as("postPlayers");

        cy.get("input[name=name").type(`${name}{enter}`);

        // Wait for the POST request (adding a new player)
        cy.wait("@postPlayers");

        // Wait for the GET request (fetching all players)
        cy.wait("@getPlayers");
    });
});
